src/backend/columnar/columnar_tableam.c

```diff
@@ -341,24 +364,34 @@ columnar_parallelscan_reinitialize(Relation rel, ParallelTableScanDesc pscan)
 static IndexFetchTableData *
 columnar_index_fetch_begin(Relation rel)
 {
-	ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-					errmsg("indexes not supported for columnar tables")));
+	Oid relfilenode = rel->rd_node.relNode;
+	if (PendingWritesInUpperTransactions(relfilenode, GetCurrentSubTransactionId()))
+	{
+		/* XXX: maybe we can just flush the data and continue */
+		elog(ERROR, "cannot read from index when there is unflushed data in "
+					"upper transactions");
+	}
+
+	FlushWriteStateForRelfilenode(relfilenode, GetCurrentSubTransactionId());
+
+	IndexFetchTableData *scan = palloc0(sizeof(IndexFetchTableData));
+	scan->rel = rel;
+	return scan;
 }
 
 
 static void
 columnar_index_fetch_reset(IndexFetchTableData *scan)
 {
-	ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-					errmsg("indexes not supported for columnar tables")));
+	/* no-op */
 }
 
 
 static void
 columnar_index_fetch_end(IndexFetchTableData *scan)
 {
-	ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-					errmsg("indexes not supported for columnar tables")));
+	columnar_index_fetch_reset(scan);
+	pfree(scan);
 }
 
 
```

Comment:
Even if `EXPLAIN` tells that we will do index-only scan, it doesn't seem to be the case since the time it takes to do "index scan" and "index-only scan" was the same for columnar tables in my experiments.

```sql
CREATE TABLE columnar_table (a INT, b int) USING columnar;
INSERT INTO columnar_table SELECT i, i*2 FROM generate_series(1000, 20000) i;
CREATE INDEX ON columnar_table (a);

-- force using index-only scan
SET columnar.enable_custom_scan TO off;
```

```sql
EXPLAIN SELECT a FROM columnar_table WHERE a = 15433;
┌─────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                           QUERY PLAN                                            │
├─────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Index Only Scan using columnar_table_a_idx on columnar_table  (cost=0.29..5.95 rows=95 width=4) │
│   Index Cond: (a = 15433)                                                                       │
└─────────────────────────────────────────────────────────────────────────────────────────────────┘
(2 rows)

SELECT a FROM columnar_table WHERE a = 15433;
┌───────┐
│   a   │
├───────┤
│ 15433 │
└───────┘
(1 row)

Time: 1.174 ms
```

```sql
-- disable index only scan to force using regular "index scan"
SET enable_indexonlyscan to OFF;

EXPLAIN SELECT a FROM columnar_table WHERE a = 15433;
┌─────────────────────────────────────────────────────────────────────────────────────────────┐
│                                         QUERY PLAN                                          │
├─────────────────────────────────────────────────────────────────────────────────────────────┤
│ Index Scan using columnar_table_a_idx on columnar_table  (cost=0.29..61.95 rows=95 width=4) │
│   Index Cond: (a = 15433)                                                                   │
└─────────────────────────────────────────────────────────────────────────────────────────────┘
(2 rows)


SELECT a FROM columnar_table WHERE a = 15433;
┌───────┐
│   a   │
├───────┤
│ 15433 │
└───────┘
(1 row)

Time: 1.281 ms
```

Made some debugging to understand the reason and what happens is:

Postgres calls `IndexOnlyNext` function to do index-only scan on columnar table. On the other hand, the [check](https://github.com/postgres/postgres/blob/master/src/backend/executor/nodeIndexonlyscan.c#L161) here always returns true for index scans on columnar tables, even if index-only scan seems possible.

Normally, what happens for heapAM tables is, above check returns false and somewhere in [here](https://github.com/postgres/postgres/blob/master/src/backend/executor/nodeIndexonlyscan.c#L199..L211) postgres directly stores the index rel tuple.

Reply:
Made some hacky changes for that b89ec736a3b2acbd98ca354d60278cb038ac26b8.